torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -408,16 +408,17 @@ ProcessGroupNCCL::ProcessGroupNCCL(
         "Invalid value for environment variable: " +
         std::string(NCCL_BLOCKING_WAIT));
   }
-
-  // If single-process single-device mode, WorkNCCL::getFuture is supported,
-  // so get a dedicated stream for each device to run FutureNCCL then callbacks.
-  // Depending on the device index of collective outputs, WorkNCCL passes
+  // If single-process single-device mode, WorkNCCL::getFuture is supported.
+  // Depending on the device index of collective outputs, WorkNCCL will pass
   // the corresponding device's then callback stream to FutureNCCL.
+  // For each device we first put a nullptr as a placeholder. Depending on the
+  // device of the NCCL collective's outputs, we later set the callback stream
+  // of the corresponding device inside ProcessGroupNCCL::getNCCLComm if not set
+  // before.
   futureNCCLCallbackStreams_.reserve(c10::cuda::device_count());
   for (int device_index = 0; device_index < c10::cuda::device_count();
        device_index++) {
-    futureNCCLCallbackStreams_.push_back(std::make_shared<at::cuda::CUDAStream>(
-        at::cuda::getStreamFromPool(device_index)));
+    futureNCCLCallbackStreams_.push_back(nullptr);
```

Comment:
Is there a reason we need to push a `nullptr` here and then push the stream later in `getNCCLComm`? Is it because we don't know the list of devices to push streams for, and it's not necessarily correct to push them for all devices?

Reply:
When we push the streams in the constructor, each rank occupies memory in all devices:
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0    673925      C   python                                       771MiB |
|    0    673926      C   python                                       771MiB |
|    1    673925      C   python                                       771MiB |
|    1    673926      C   python                                       771MiB |
|    2    673925      C   python                                       771MiB |
|    2    673926      C   python                                       771MiB |
|    3    673925      C   python                                       771MiB |
|    3    673926      C   python                                       771MiB |
|    4    673925      C   python                                       771MiB |
|    4    673926      C   python                                       771MiB |
|    5    673925      C   python                                       771MiB |
|    5    673926      C   python                                       771MiB |
|    6    673925      C   python                                       771MiB |
|    6    673926      C   python                                       771MiB |
|    7    673925      C   python                                       707MiB |
|    7    673926      C   python                                       623MiB |
+-----------------------------------------------------------------------------+
``` 
That's actually a bit strange because if I just push default streams for all devices in the `ProcessGroupNCCL` constructor, it doesn't allocate additional memory on different devices. Maybe creating a pool when calling `getStreamFromPool` is the problem?

Reply:
In the constructor we also don't know which rank occupies which GPU right? I tried to print `default_device` and it was printing GPU0 for all ranks.